src/pip/_internal/operations/prepare.py

```diff
@@ -559,24 +576,33 @@ def _prepare_linked_requirement(
             assert link.is_file
             # We need to verify hashes, and we have found the requirement in the cache
             # of locally built wheels.
-            if (
-                isinstance(req.download_info.info, ArchiveInfo)
-                and req.download_info.info.hashes
-                and hashes.has_one_of(req.download_info.info.hashes)
-            ):
-                # At this point we know the requirement was built from a hashable source
-                # artifact, and we verified that the cache entry's hash of the original
-                # artifact matches one of the hashes we expect. We don't verify hashes
-                # against the cached wheel, because the wheel is not the original.
+            if isinstance(req.download_info.info, ArchiveInfo):
+                if req.download_info.info.hashes and hashes.has_one_of(
+                    req.download_info.info.hashes
+                ):
+                    # At this point we know the requirement was built from a hashable
+                    # source artifact, and we verified that the cache entry's hash of
+                    # the original artifact matches one of the hashes we expect. We
+                    # don't verify hashes against the cached wheel, because the wheel is
+                    # not the original.
+                    hashes = None
+                else:
+                    logger.warning(
+                        "The hashes of the source archive found in cache entry "
+                        "don't match, ignoring cached built wheel "
+                        "and re-downloading source."
+                    )
+                    req.link = req.cached_wheel_source_link
+                    link = req.link
+            elif isinstance(req.download_info.info, VcsInfo):
+                if not req.user_supplied:
+                    raise DependencyVcsHashNotSupported()
```

Comment:
I wonder if this would break anything. How can this branch be reached?

Reply:
This would raise in require-hashes mode only, if you have a dependency as a git direct URL that is obtained from cache.

I think these should not be considered hash-valid unless the commit hash has been provided by the user, either as a requirement file or a constraint.

This PR still need work. Next step is adding a lot of test coverage. Then possibly some refactoring.

Reply:
If an immutable VCS URL is both user-supplied and a dependency, would this cause the latter specifier break installation?

Reply:
It should not, because when the dependency is examined by the resolver it is "merged" with the user-supplied one? That's definitely a test case to consider.